nixos/paperless: download NLTK data

puppe opened the pull request:
###### Description of changes

Since version 1.10.0 paperless-ngx depends on the NLTK library which is
used to pre-process data for machine learning. NLTK needs certain
data for stemming, stopword removal etc. This data has to be downloaded
first. This pull request introduces a new systemd service that does the
downloading.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [ ] ~~[23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)~~
  - [ ] ~~(Package updates) Added a release notes entry if the change is major or breaking~~
  - [ ] ~~(Module updates) Added a release notes entry if the change is significant~~
  - [ ] ~~(Module addition) Added a release notes entry if adding a new NixOS module~~
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

puppe committed:
nixos/paperless: apply nixpkgs-fmt

puppe committed:
nixos/paperless: download NLTK data

Since version 1.10.0 paperless-ngx depends on the NLTK library which is
used to pre-process data for machine learning. NLTK needs certain
data for stemming, stopword removal etc. This data has to be downloaded
first. This commit introduces a new systemd service that does the
downloading.

puppe committed:
paperless-ngx: update homepage URL

gador reviewed (approved):
Tested this PR in a VM. Works just fine (didn't see any errors in the logs, NLTK data is downloaded, documents are consumed just fine). Tests pass, also

erikarvstedt reviewed (commented):
Thanks! I will review this in the next few days.

mweinelt commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -291,6 +295,72 @@ in
+      before = [ "paperless-scheduler.service" ];
+      script = let pythonWithNltk = pkg.python.withPackages (ps: [ ps.nltk ]); in
+        ''
+          test -d "${nltkDir}" || mkdir "${nltkDir}"
```
```suggestion
          mkdir -p "${nltkDir}"
```

Should accomplish the same, albeit simpler?

puppe replied:
Your comment prompted me to have another look at this. I have updated my pull request. The NLTK directory is now created via systemd.tmpfiles.rules, like the other directories.

puppe commented:
I have updated my pull request. See #https://github.com/NixOS/nixpkgs/pull/215750#discussion_r1112306930.

erikarvstedt reviewed (changes requested):
[Fixup](https://github.com/erikarvstedt/nixpkgs/commits/fix-paperless-ea).

- Fetch with:
  ```bash
  git fetch https://github.com/erikarvstedt/nixpkgs fix-paperless-ea
  ```

- Description:
  Use `defaultServiceConfig` for systemd hardening, which uses `TemporaryFileSystem` instead of `ProtectSystem` to restrict filesystem access.
  This method allows setting `paperless.user` to an existing real user, which is a use case required by the original author (@Flakebi).
  This was broken by the previous hardening setting.

### Systemd sandboxing behavior
@puppe, you added this code comment regarding service sandboxing:
> but even with the '+' prefix, some restrictions were still applied to the download script.
> Those restrictions prevented the script from downloading any data. This might be a bug in systemd

This is expected behavior. The '+' prefix just sets the process user to `root`, but all namespace related settings are unchanged.
Anyways, using a dedicated service is cleaner, particularly in case of offline deployments.

erikarvstedt commented:
@puppe, is there any way to test if the NLTK data is picked up and used during classification?
Did you manually check if this is the case?

puppe commented:
>     * Description:
>       Use `defaultServiceConfig` for systemd hardening, which uses `TemporaryFileSystem` instead of `ProtectSystem` to restrict filesystem access.

Thanks! I have incorporated your changes. I would have preferred to do it that way anyway to be consistent with the other services. But to be honest, I wasn‘t able to figure out which paths to include in `BindReadOnlyPaths` to make it work. So, I fell back to do it the way I usually harden services (doing what `systemd-analyze security` tells me).

> > but even with the '+' prefix, some restrictions were still applied to the download script.
> > Those restrictions prevented the script from downloading any data. This might be a bug in systemd
> 
> This is expected behavior. The '+' prefix just sets the process user to `root`, but all namespace related settings are unchanged. It's possible to run the downloader in `preStart` by switching to the root network namespace, like so:

You‘re right. I had another look at `man systemd.service`. Quote (emphasis mine):

> If the executable path is prefixed with "+" then the process is executed with full privileges. In this mode privilege restrictions configured with User=, Group=, CapabilityBoundingSet= or the various *file system* namespacing options (such as PrivateDevices=, PrivateTmp=) are not applied to the invoked command line […].

The `+` prefix does not affect *network* namespacing so it‘s not a bug in systemd (but still counterintuitive IMO).

In case anyone is wondering. Yes I force-pushed three times in the last few minutes. At first, I had rebased the pull request onto the current master branch which was a bad idea. It made comparing the old version to the new version ususable because of all the changes in the master branch. So, I force-pushed the old version again. And then I force-pushed the new version with your changes (but without rebasing onto the current master branch).

puppe commented:
> @puppe, is there a way to test if the NLTK data is picked up and used during classification? Did you manually check if this is the case?

I have checked manually. If the data is absent, you get an error like the following in the log when the classification task is started:

```
[2023-02-23 15:57:43,737] [WARNING] [paperless.tasks] Classifier error:
**********************************************************************
  Resource stopwords not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('stopwords')

  For more information see: https://www.nltk.org/data.html

  Attempted to load corpora/stopwords

  Searched in:
    - '/usr/share/nltk_data'
**********************************************************************
```

With my changes, you don‘t get an error and classification works.

puppe commented:
And BTW, it will not re-download the data if there is no update:

```
Feb 23 16:04:11 giles paperless-download-nltk-data-start[1039731]: [nltk_data] Downloading package punkt to /var/lib/paperless/nltk...
Feb 23 16:04:11 giles paperless-download-nltk-data-start[1039731]: [nltk_data]   Package punkt is already up-to-date!
Feb 23 16:04:11 giles paperless-download-nltk-data-start[1039731]: [nltk_data] Downloading package snowball_data to
Feb 23 16:04:11 giles paperless-download-nltk-data-start[1039731]: [nltk_data]     /var/lib/paperless/nltk...
Feb 23 16:04:11 giles paperless-download-nltk-data-start[1039731]: [nltk_data]   Package snowball_data is already up-to-date!
Feb 23 16:04:11 giles paperless-download-nltk-data-start[1039731]: [nltk_data] Downloading package stopwords to
Feb 23 16:04:11 giles paperless-download-nltk-data-start[1039731]: [nltk_data]     /var/lib/paperless/nltk...
Feb 23 16:04:11 giles paperless-download-nltk-data-start[1039731]: [nltk_data]   Package stopwords is already up-to-date!
```

erikarvstedt reviewed (approved):
*removed*

erikarvstedt reviewed (changes requested):
[Fixup](https://github.com/erikarvstedt/nixpkgs/commits/fix-paperless-ea-2)

puppe commented:
> [Fixup](https://github.com/erikarvstedt/nixpkgs/commits/fix-paperless-ea-2)

As you suggested, I have changed the directory for NLTK data to /var/cache/paperless/nltk. The directory /var/cache/paperless will be created automatically because of `CacheDirectory = [ "paperless ]";`. NLTK will create the nltk subdirectory automatically.

erikarvstedt reviewed (approved):
*Removed*

erikarvstedt commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -291,6 +296,31 @@ in
+        User = cfg.user;
+        Type = "oneshot";
+        # Enable internet access
+        PrivateNetwork = false;
```
Restrict the downloader to only access the cache dir.
```suggestion
        PrivateNetwork = false;
        BindPaths = [];
```

erikarvstedt commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -47,6 +51,7 @@ let
       cfg.dataDir
       cfg.mediaDir
     ];
+    CacheDirectory = [ "paperless" ];
```
```suggestion
    CacheDirectory = "paperless";
```

puppe commented:
@erikarvstedt Your latest suggestions are now included.

erikarvstedt reviewed (approved):
- [x] `nixosTests.paperless` succeeds on `x86_64-linux`
- [x] I've manually checked that classification using the NLTK data works correctly

puppe commented:
Three reviewers have approved this pull request. What is the way forward now?

felschr commented:
The new `paperless-download-nltk-data.service` fails for me after upgrading from NixOS 22.11 to 23.05:

```console
❯ journalctl -u paperless-download-nltk-data.service
May 31 19:11:10 home-server systemd[1]: Starting paperless-download-nltk-data.service...
May 31 19:11:13 home-server python[1311172]: /nix/store/95cxzy2hpizr23343b8bskl4yacf4b3l-python3-3.10.11/lib/python3.10/runpy.py:126: RuntimeWarning: 'nltk.downloader' found in sys.modules after import of package 'nltk', but prior to execution of 'nltk.downloader'; this may result i>
May 31 19:11:13 home-server python[1311172]:   warn(RuntimeWarning(msg))
May 31 19:11:14 home-server python[1311172]: [nltk_data] Error loading punkt: <urlopen error [Errno -3] Temporary
May 31 19:11:14 home-server python[1311172]: [nltk_data]     failure in name resolution>
May 31 19:11:14 home-server python[1311172]: Error installing package. Retry? [n/y/e]
May 31 19:11:14 home-server python[1311172]: Traceback (most recent call last):
May 31 19:11:14 home-server python[1311172]:   File "/nix/store/95cxzy2hpizr23343b8bskl4yacf4b3l-python3-3.10.11/lib/python3.10/runpy.py", line 196, in _run_module_as_main
May 31 19:11:14 home-server python[1311172]:     return _run_code(code, main_globals, None,
May 31 19:11:14 home-server python[1311172]:   File "/nix/store/95cxzy2hpizr23343b8bskl4yacf4b3l-python3-3.10.11/lib/python3.10/runpy.py", line 86, in _run_code
May 31 19:11:14 home-server python[1311172]:     exec(code, run_globals)
May 31 19:11:14 home-server python[1311172]:   File "/nix/store/9b4lph7y3pqd5v5nzfm5bv8f8bm6afn0-python3-3.10.11-env/lib/python3.10/site-packages/nltk/downloader.py", line 2544, in <module>
May 31 19:11:14 home-server python[1311172]:     rv = downloader.download(
May 31 19:11:14 home-server python[1311172]:   File "/nix/store/9b4lph7y3pqd5v5nzfm5bv8f8bm6afn0-python3-3.10.11-env/lib/python3.10/site-packages/nltk/downloader.py", line 788, in download
May 31 19:11:14 home-server python[1311172]:     choice = input().strip()
May 31 19:11:14 home-server python[1311172]: EOFError: EOF when reading a line
May 31 19:11:14 home-server systemd[1]: paperless-download-nltk-data.service: Main process exited, code=exited, status=1/FAILURE
May 31 19:11:14 home-server systemd[1]: paperless-download-nltk-data.service: Failed with result 'exit-code'.
May 31 19:11:14 home-server systemd[1]: Failed to start paperless-download-nltk-data.service.
May 31 19:11:14 home-server systemd[1]: paperless-download-nltk-data.service: Triggering OnFailure= dependencies.
May 31 19:11:14 home-server systemd[1]: paperless-download-nltk-data.service: Consumed 890ms CPU time, received 0B IP traffic, sent 1.5K IP traffic.
```

puppe commented:
Does this happen everytime? What happens when you do `systemctl start paperless-download-nltk-data.service`? What happens when you do `dig raw.githubusercontent.com` or `nslookup raw.githubusercontent.com`?
